File: backend/app/schemas/alert.py
"""
backend/app/schemas/alert.py
Pydantic schemas for metric alert rules and evaluation results.
"""
from typing import Optional, Literal
from datetime import datetime
from pydantic import BaseModel, Field


class AlertRuleCreate(BaseModel):
    name: str
    dataset_id: str
    metric_column: str
    aggregate_fn: Literal["AVG", "SUM", "MAX", "MIN", "COUNT"] = "AVG"
    operator: Literal[">", "<", ">=", "<=", "=="]
    threshold: float
    severity: Literal["info", "warning", "critical"] = "warning"
    description: Optional[str] = None


class AlertRuleResponse(BaseModel):
    id: str
    name: str
    dataset_id: str
    metric_column: str
    aggregate_fn: str
    operator: str
    threshold: float
    severity: str
    is_active: bool
    triggered: bool
    last_value: Optional[float] = None
    last_checked_at: Optional[datetime] = None
    last_triggered_at: Optional[datetime] = None
    description: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class AlertEvalResult(BaseModel):
    rule_id: str
    rule_name: str
    current_value: float
    threshold: float
    operator: str
    triggered: bool
    severity: str
    message: str

File: frontend/types/alert.ts
// frontend/types/alert.ts
export interface AlertRuleCreate {
  name: string;
  dataset_id: string;
  metric_column: string;
  aggregate_fn: "AVG" | "SUM" | "MAX" | "MIN" | "COUNT";
  operator: ">" | "<" | ">=" | "<=" | "==";
  threshold: number;
  severity: "info" | "warning" | "critical";
  description?: string;
}

export interface AlertRule {
  id: string;
  name: string;
  dataset_id: string;
  metric_column: string;
  aggregate_fn: string;
  operator: string;
  threshold: number;
  severity: string;
  is_active: boolean;
  triggered: boolean;
  last_value?: number | null;
  last_checked_at?: string | null;
  last_triggered_at?: string | null;
  description?: string | null;
  created_at: string;
}

export interface AlertEvalResult {
  rule_id: string;
  rule_name: string;
  current_value: number;
  threshold: number;
  operator: string;
  triggered: boolean;
  severity: string;
  message: string;
}
